File: scripts/verify_zones.ts
import { calculatePaceZones, calculateHRZones, resolvePaceTarget, resolveHRTarget, formatPace } from '../src/lib/training/zones';

function runVerification() {
    console.log("Starting Zone Verification...");

    // 1. Test Profile
    const thresholdPace = 240; // 4:00/km
    const lthr = 170; // bpm

    console.log(`\nProfile: TP=${formatPace(thresholdPace)}/km, LTHR=${lthr} bpm`);

    // 2. Verify Pace Zones
    console.log("\n--- Pace Zones ---");
    const paceZones = calculatePaceZones(thresholdPace);
    console.log("Z1 (Easy):", formatPace(paceZones.Z1.min), "-", formatPace(paceZones.Z1.max));
    console.log("Z4 (Threshold):", formatPace(paceZones.Z4.min), "-", formatPace(paceZones.Z4.max));

    // Check Z4 Logic: 1.00 - 1.05 * TP. 
    // Lower Bound (Fastest) = 1.00 * 240 = 240 (4:00)
    // Upper Bound (Slowest) = 1.05 * 240 = 252 (4:12)
    if (Math.abs(paceZones.Z4.min - 240) < 1 && Math.abs(paceZones.Z4.max - 252) < 1) {
        console.log("✅ Z4 Pace Calculation Correct");
    } else {
        console.error("❌ Z4 Pace Calculation Failed");
    }

    // 3. Verify HR Zones
    console.log("\n--- HR Zones ---");
    const hrZones = calculateHRZones(lthr);
    console.log("Z4 (Threshold):", hrZones.Z4.min, "-", hrZones.Z4.max);
    // Check Z4 Logic: 0.95 - 0.99 * LTHR
    // Min = 170 * 0.95 = 161.5 -> 162
    // Max = 170 * 0.99 = 168.3 -> 168
    if (hrZones.Z4.min === 162 && hrZones.Z4.max === 168) {
        console.log("✅ Z4 HR Calculation Correct");
    } else {
        console.error("❌ Z4 HR Calculation Failed");
    }

    // 4. Verify Target Resolution
    console.log("\n--- Target Resolution ---");
    // Z4 Middle (0.5)
    const z4Target = resolvePaceTarget(thresholdPace, 'Z4', 0.5);
    // Range: 240 - 252. Diff = 12. 0.5 * 12 = 6. Max(252) - 6 = 246.
    console.log("Z4 @ 0.5 Target:", formatPace(z4Target));

    if (Math.abs(z4Target - 246) < 1) {
        console.log("✅ Z4 Target Resolution Correct");
    } else {
        console.error("❌ Z4 Target Resolution Failed. Got:", z4Target);
    }

    // 5. Verify Update Behavior (Change TP)
    console.log("\n--- Update Behavior (New TP) ---");
    const newTP = 300; // 5:00/km
    const newZ4Target = resolvePaceTarget(newTP, 'Z4', 0.5);
    console.log(`New TP: ${formatPace(newTP)}/km`);
    console.log("New Z4 @ 0.5 Target:", formatPace(newZ4Target));

    // Range: 300 - 315. Diff 15. 0.5 * 15 = 7.5. 315 - 7.5 = 307.5.
    if (Math.abs(newZ4Target - 307.5) < 1) {
        console.log("✅ Dynamic Update Verified");
    } else {
        console.error("❌ Dynamic Update Failed");
    }
}

runVerification();

File: scripts/verify_zones.py
import math

def calculate_pace_zones(threshold_pace):
    return {
        'Z1': {'min': threshold_pace * 1.30, 'max': threshold_pace * 1.45},
        'Z2': {'min': threshold_pace * 1.15, 'max': threshold_pace * 1.30},
        'Z3': {'min': threshold_pace * 1.05, 'max': threshold_pace * 1.15},
        'Z4': {'min': threshold_pace * 1.00, 'max': threshold_pace * 1.05},
        'Z5': {'min': threshold_pace * 0.90, 'max': threshold_pace * 1.00},
    }

def calculate_hr_zones(lthr):
    return {
        'Z1': {'min': 0, 'max': round(lthr * 0.85) - 1},
        'Z2': {'min': round(lthr * 0.85), 'max': round(lthr * 0.89)},
        'Z3': {'min': round(lthr * 0.90), 'max': round(lthr * 0.94)},
        'Z4': {'min': round(lthr * 0.95), 'max': round(lthr * 0.99)},
        'Z5': {'min': round(lthr * 1.00), 'max': 220},
    }

def format_pace(seconds):
    mins = math.floor(seconds / 60)
    secs = round(seconds % 60)
    return f"{mins}:{secs:02d}"

def resolve_pace_target(threshold_pace, zone, position):
    zones = calculate_pace_zones(threshold_pace)
    rng = zones[zone]
    # Easiest (Pos 0) is MAX seconds (slower)
    return rng['max'] - (position * (rng['max'] - rng['min']))

def run_verification():
    print("Starting Zone Verification...")

    # 1. Test Profile
    threshold_pace = 240 # 4:00/km
    lthr = 170 # bpm

    print(f"\nProfile: TP={format_pace(threshold_pace)}/km, LTHR={lthr} bpm")

    # 2. Verify Pace Zones
    print("\n--- Pace Zones ---")
    pace_zones = calculate_pace_zones(threshold_pace)
    print(f"Z1 (Easy): {format_pace(pace_zones['Z1']['min'])} - {format_pace(pace_zones['Z1']['max'])}")
    print(f"Z4 (Threshold): {format_pace(pace_zones['Z4']['min'])} - {format_pace(pace_zones['Z4']['max'])}")
    
    if abs(pace_zones['Z4']['min'] - 240) < 1 and abs(pace_zones['Z4']['max'] - 252) < 1:
        print("✅ Z4 Pace Calculation Correct")
    else:
        print("❌ Z4 Pace Calculation Failed")

    # 3. Verify HR Zones
    print("\n--- HR Zones ---")
    hr_zones = calculate_hr_zones(lthr)
    print(f"Z4 (Threshold): {hr_zones['Z4']['min']} - {hr_zones['Z4']['max']}")
    
    if hr_zones['Z4']['min'] == 162 and hr_zones['Z4']['max'] == 168:
        print("✅ Z4 HR Calculation Correct")
    else:
        print("❌ Z4 HR Calculation Failed")

    # 4. Verify Target Resolution
    print("\n--- Target Resolution ---")
    z4_target = resolve_pace_target(threshold_pace, 'Z4', 0.5)
    print(f"Z4 @ 0.5 Target: {format_pace(z4_target)}")
    
    if abs(z4_target - 246) < 1:
        print("✅ Z4 Target Resolution Correct")
    else:
        print(f"❌ Z4 Target Resolution Failed. Got: {z4_target}")

    # 5. Verify Update Behavior
    print("\n--- Update Behavior (New TP) ---")
    new_tp = 300 # 5:00/km
    new_z4_target = resolve_pace_target(new_tp, 'Z4', 0.5)
    print(f"New TP: {format_pace(new_tp)}/km")
    print(f"New Z4 @ 0.5 Target: {format_pace(new_z4_target)}")
    
    if abs(new_z4_target - 307.5) < 1:
         print("✅ Dynamic Update Verified")
    else:
         print("❌ Dynamic Update Failed")

if __name__ == "__main__":
    run_verification()
